bpf: avoid encrypt_key map lookup if IPsec is disabled

tklauser opened the pull request:
Avoid an unnecessary map lookup for encrypt_key in the bpf_lxc program in case IPsec is disabled. Also fix warnings/errors when building with LLVM 14. See individual commit messages for details.

tklauser committed:
bpf: fix -Wunused-but-set-variable errors when building with LLVM 14

Building the BPF datapath with LLVM 14 leads to the following errors:

    bpf_lxc.c:101:16: error: variable 'daddr' set but not used [-Werror,-Wunused-but-set-variable]
            union v6addr *daddr, orig_dip;
                          ^
    bpf_lxc.c:103:7: error: variable 'encrypt_key' set but not used [-Werror,-Wunused-but-set-variable]
            __u8 encrypt_key = 0;
                 ^
    bpf_lxc.c:102:8: error: variable 'tunnel_endpoint' set but not used [-Werror,-Wunused-but-set-variable]
            __u32 tunnel_endpoint = 0;
                  ^
    bpf_lxc.c:526:7: error: variable 'encrypt_key' set but not used [-Werror,-Wunused-but-set-variable]
            __u8 encrypt_key = 0;
                 ^
    bpf_lxc.c:525:8: error: variable 'tunnel_endpoint' set but not used [-Werror,-Wunused-but-set-variable]
            __u32 tunnel_endpoint = 0;
                  ^

These are normally warnings, but errors in this case due to the use of
-Werror when compiling Cilium's bpf programs. Fix these by marking the
affected variables as __maybe_unused.

Signed-off-by: Tobias Klauser <tobias@cilium.io>

pchaigno commented on bpf/bpf_lxc.c:
```diff
@@ -196,7 +196,9 @@ static __always_inline int ipv6_l3_from_lxc(struct __ctx_buff *ctx,
 			tunnel_endpoint = info->tunnel_endpoint;
+#ifdef ENABLE_IPSEC
 			encrypt_key = get_min_encrypt_key(info->key);
+#endif
```
Should we move the preprocessor guard inside `get_min_encrypt_key` to avoid having too many preprocessor directives in this function/file? We already have many and it doesn't help readability IMO.

tklauser replied:
Good point, will move the preprocessor guard into `get_min_encrypt_key`.

pchaigno reviewed (approved):
Looks good to me; one nit below.

Did you also try to load the programs compiled with LLVM 14?

tklauser committed:
bpf: avoid encrypt_key map lookup if IPsec is disabled

In the bpf_lxc program's functions ipv6_l3_from_lxc and
handle_ipv4_from_lxc, currently encrypt_key is always looked up in the
encrypt map, regardless of whether IPsec is enabled or not. However, its
value is only actually used when IPsec is enabled. Thus, the call can
be avoid when IPsec is disabled.

This also slightly reduces program size if !defined(ENABLE_IPSEC).

Signed-off-by: Tobias Klauser <tobias@cilium.io>
